source/server/guarddog_impl.cc

```diff
@@ -83,25 +127,47 @@ void GuardDogImpl::step() {
           watched_dog->megamiss_counter_.inc();
           watched_dog->last_alert_time_ = ltt;
           watched_dog->megamiss_alerted_ = true;
+          mega_miss_threads.emplace_back(tid, ltt);
         }
       }
       if (killEnabled() && delta > kill_timeout_) {
+        // Runs all of the registered kill callbacks the order thery were registered.
+        for(GuardDogActionCb& cb : event_to_callbacks_[WatchDogEvent::Watchdog_WatchdogAction_WatchdogEvent_KILL]) {
+          cb(WatchDogEvent::Watchdog_WatchdogAction_WatchdogEvent_KILL, {{tid, ltt}}, now);
+        }
         PANIC(fmt::format("GuardDog: one thread ({}) stuck for more than watchdog_kill_timeout",
                           watched_dog->dog_->threadId().debugString()));
       }
       if (multikillEnabled() && delta > multi_kill_timeout_) {
-        if (seen_one_multi_timeout) {
+        multi_kill_threads.emplace_back(tid, ltt);
 
-          PANIC(fmt::format(
-              "GuardDog: multiple threads ({},...) stuck for more than watchdog_multikill_timeout",
-              watched_dog->dog_->threadId().debugString()));
-        } else {
-          seen_one_multi_timeout = true;
+        if (multi_kill_threads.size() >= required_for_multi_kill) {
+          // Runs all of the registered multikill callbacks the order thery were registered.
+          for(GuardDogActionCb& cb : event_to_callbacks_[WatchDogEvent::Watchdog_WatchdogAction_WatchdogEvent_MULTIKILL]) {
```

Comment:
operator[] is a mutation operation.  I think you want to use find.

Also, could you add a helper method to invoke the relevant watchdog actions and avoid having to duplicate the iteration code writting out the event name in 2 places?  something like:
invokeGuardDogActions(event, thread_ids, now)